### ISSUE TITLE
If have a 140-gene list, then want to scale the gene panel to 500

### ISSUE THREAD

**Issue by KunHHE** [theislab/spapros#389]
Hi @LouisK92, I'm so sorry. I'am new python player, but very interested in running this spapros. I know the case if more than 150gene panel needed, multiple selection is necessary.
I saw the tutorial two scenarios: We preselected 40 genes and want to add 10 (adds up to 50 in total); We preselected 40 genes and want to add 100 (adds up to 140 in total). If I have a 140-gene list, then I want to scale the gene panel to 500, 

Can you guide me on how to best perform the selection? We add 140 panel list and select additional 10 gene as the first selection? Then perform the left rounds of selection? 

# list of pre selected genes
preselected_genes = [
**I can list 140 gene panel list here**
]
# Get `n_pca_genes` to obtain the top 20 pca genes that are not already in preselected_genes
tmp = sp.se.select_pca_genes(adata, n=adata.n_vars, inplace=False)["selection_ranking"]
n_pca_genes = tmp.loc[~tmp.index.isin(preselected_genes)].sort_values().iloc[:20].max().astype(int)

# create an instance of the ProbesetSelector class
selector = sp.se.ProbesetSelector(
    adata,
    n=50,
    n_pca_genes=n_pca_genes,
    preselected_genes=preselected_genes,
    celltype_key="celltype",
    verbosity=1,
    save_dir=None,
)
# select probe set
selector.select_probeset()

**Comment by LouisK92** [comment]
Hi @KunHHE - no worries! Thanks for your interest, please keep the questions coming.

For your scenario I would go as follows:
1. For the first selection I would ignore the preselected genes. With this selection the goal is to maximise robustness for cell type recovery. If there are already some good marker genes in the preselection we don't want the algorithm to reuse those. Instead additional markers should be selected.
2. The second selection will take the pre selected genes into account but not the one from the first selection. We ignore the markers of the first selection to fill up signals for the cell types that weren't captured with the preselection yet and increase robustness on those. 

Sounds a bit complicated, sorry, hope it makes sense. 

So the selection to run would look like this:

```python
import numpy as np
import pandas as pd
import scanpy as sc
import spapros as sp


# Taking the data example of the tutorials
adata = sc.datasets.pbmc3k()
adata_tmp = sc.datasets.pbmc3k_processed()
adata = adata[adata_tmp.obs_names,adata_tmp.var_names]
adata.obs['celltype'] = adata_tmp.obs['louvain']
adata.obsm['X_umap'] = adata_tmp.obsm['X_umap']
del adata_tmp
sc.pp.normalize_total(adata)
sc.pp.log1p(adata)
sc.pp.highly_variable_genes(adata,flavor="cell_ranger",n_top_genes=1000)


# List of pre selected genes (here just randomly selected 140 genes)
preselected_genes = list(np.random.choice(adata.var_names, 140, replace=False))

# Selection

# First run
selector = sp.se.ProbesetSelector(
    adata[:,[v for v in adata.var_names if v not in preselected_genes]], # NOTE: remove preselected genes in 1st run
    n=None, 
    n_pca_genes=0, # NOTE: no PCA in 1st run
    celltype_key="celltype",
    verbosity=1,
    save_dir=None,
)
selector.select_probeset()

# Save genes of the first selection
first_selection = selector.probeset[selector.probeset["selection"]].index.tolist()
n_genes_first_selection = len(first_selection)
print(n_genes_first_selection)


# Second run
selector2 = sp.se.ProbesetSelector(
    adata[:,[v for v in adata.var_names if v not in first_selection]], # NOTE: ignore genes of 1st run in 2nd run
    n=500 - n_genes_first_selection, # NOTE: select 500 genes in total
    preselected_genes=preselected_genes, # NOTE: include preselected genes in 2nd run
    n_pca_genes=100, # NOTE: PCA selection in 2nd run
    celltype_key="celltype",
    verbosity=1,
    save_dir=None,
)

# select probe set
selector2.select_probeset()

# Concatenate the two selections
selector2.probeset['gene_nr'] += n_genes_first_selection
probeset = pd.concat([selector.probeset.loc[selector.probeset["selection"]], selector2.probeset])

```

**Comment by KunHHE** [comment]
Hi @LouisK92, appreciate your quick response! Kind of confused, so in the case of the preselected gene list in hand, when you say # #List of pre selected genes (here just randomly selected 140 genes) run: preselected_genes = list(np.random.choice(adata.var_names, 140, replace=False)). Should I list my 140 gene list here? Because in the script that you shared above, this is the only place to list interested gene list? For example: 
# list of pre selected genes
preselected_genes = ['PF4', 'HLA-DPB1', 'FCGR3A', 'GZMB', 'CCL5', 'S100A8', 'IL32', 'HLA-DQA1', 'NKG7', 'AIF1', 'CD79A', 'LTB', 'TYROBP','HLA-DMA', 'GZMK', 'HLA-DRB1', 'FCN1', 'S100A11', 'GNLY', 'GZMH', 'CST3', 'RBM39', 'HLA-DPA1', 'LST1', 'LGALS2',
 'FCER1A', 'MS4A1', 'TRAF3IP3', 'PRF1', 'LGALS1', 'PRDX1', 'ATP5O', 'C1orf162', 'CTSS', 'SELL', 'CTSW', 'GIMAP4', 'LINC00926', 'MS4A6A', 'GSTP1'.........140 gene list] 

Then how should I modify the script? list(np.random.choice(adata.var_names, 140, replace=False)). Should I do "preselected_genes = list(np.random.choice([
    'PF4', 'HLA-DPB1', 'FCGR3A', 'GZMB', 'CCL5', 'S100A8', 'IL32', 'HLA-DQA1', 
    'NKG7', 'AIF1', 'CD79A', 'LTB', 'TYROBP', 'HLA-DMA', 'GZMK', 'HLA-DRB1', 
    'FCN1', 'S100A11', 'GNLY', 'GZMH', 'CST3', 'RBM39', 'HLA-DPA1', 'LST1', 
    'LGALS2', 'FCER1A', 'MS4A1', 'TRAF3IP3', 'PRF1', 'LGALS1', 'PRDX1', 'ATP5O', 
    'C1orf162', 'CTSS', 'SELL', 'CTSW', 'GIMAP4', 'LINC00926', 'MS4A6A', 'GSTP1'
], 140, replace=False))"? Please correct me if I am wrong and misunderstand. Thanks again!!!

**Comment by KunHHE** [comment]
I guess I should do adata.var_names=[
'PF4', 'HLA-DPB1', 'FCGR3A', 'GZMB', 'CCL5', 'S100A8', 'IL32', 'HLA-DQA1',
'NKG7', 'AIF1', 'CD79A', 'LTB', 'TYROBP', 'HLA-DMA', 'GZMK', 'HLA-DRB1',
'FCN1', 'S100A11', 'GNLY', 'GZMH', 'CST3', 'RBM39', 'HLA-DPA1', 'LST1',
'LGALS2', 'FCER1A', 'MS4A1', 'TRAF3IP3', 'PRF1', 'LGALS1', 'PRDX1', 'ATP5O',
'C1orf162', 'CTSS', 'SELL', 'CTSW', 'GIMAP4', 'LINC00926', 'MS4A6A', 'GSTP1', .....140 genes
]. Then do preselected_genes = list(np.random.choice(adata.var_names, 140, replace=False))?

**Comment by KunHHE** [comment]
I guess I should do This? : 
# List of pre selected genes
preselected_genes = [
    'Mmrn1', 'Lgmn', 'Ms4a4c', 'Tyrobp', 'Prss23', 'H2-DMb1', 'Cd3g', 'Napsa', 'Apoe', 'Cdo1', 'Ear2', 'Hilpda', 'Pcolce', 'H2-DMb2', 'Pglyrp1', 'Scd1', 'Cd3d', 'Basp1', 'Ednrb', 'Flt4', 'Ccr2', 'Aldh1a1', 'Ccl5', 'Vwf', 'Ptprb', 'Mgp', 'H2-DMa', 'Plac8', 'Gngt2', 'Bcl2a1d', 'Ppp1r14a', 'Klrd1', 'Ccr7', 'Capg', 'Rogdi', 'Sod3', 'Inmt', 'Lyz2', 'Emb', 'Septin4', 'Mgll', 'Klrb1c', 'Ly86', 'Nkg7', 'Wfdc17', 'Limd2', 'Id2', 'Ltb', 'Gyg', 'Cadm1', 'Ccl11', 'Avpi1', 'Tinagl1', 'Rnase4', 'AU021092', 'Cd24a', 'Itgal', 'Sell', 'Reln', 'Cebpb', 'Ccnd2', 'Kit', 'Gja4', 'Stmn1', 'Ms4a6c', 'Fcer1g', 'Selenom', 'H2-Oa', 'Car4', 'Rtp4', 'Lrg1', 'Rab7b', 'Eno3', 'Ms4a6b', 'Tspan8', 'Fxyd6', 'Septin3', 'Cxcl12', 'Lxn', 'Nr4a1', 'Lpl', 'Lcn2', 'Sparcl1', 'H2-Ob', 'S1pr4', 'Mfap2', 'Wfdc1', 'Bcl2a1b', 'Edn1', 'Tgtp2', 'Slc7a10', 'Igtp', 'Gmfg', 'Rarres2', 'Hp', 'Chchd10', 'Cdh13', 'Mzb1', 'Csf3r', 'Hcar2', 'Cox6a2', 'Higd1b', 'Ccr8', 'Wnt3a', 'Cd37', 'Ptprcap', 'Samsn1', 'Sftpd', 'S100a8', 'Rgs2', 'Cbr2', 'Ager', 'Ift43', 'Cxcr6', 'Itk', 'Akip1', 'Lax1', 'Kcnk3', 'Dnajc12', 'Pacsin1', 'Ms4a7', 'Car2', 'Ccl17', 'Lat', 'Gpr182', 'Xcr1', 'Oasl2', 'G0s2', 'Ifit3', 'Kdr', 'Gcsam', 'Grina', 'Vegfc', 'Treml4', 'Scube2', 'Ifi35', 'Ifi30', 'Hey1', 'Rac2', 'AA467197'......140 genes
]
preselected_genes = list(np.random.choice(adata_00.var_names, 140, replace=False)). Thanks!!!

**Comment by LouisK92** [comment]
Yes exactly the random selection that I used in the example was just a placeholder for your list of 140 preselected genes.

**Comment by KunHHE** [comment]
Hi @LouisK92 Thanks for your patience! So I successfully performed 2 rounds of selections including 126 pre-defined gene lists. But based on the script using selector + selector2.plot_gene_overlap(), it looks like the function using "selector + selector2.plot_gene_overlap()" only showed the second 380 genes but the first selection 120 genes were not pooled together (if I use selector.plot_gene_overlap(), it shows 120 genes for the first round of selection.

1. So my question is the spapros cannot show plot_gene_overlap into one plot figure including multiple rounds of selection?
2. The first selection is trying to recover all the cell types, so I think here it's a good way to do a Dotplot by using "sp.pl.masked_dotplot(adata_00, selector)",  for the first selection? I used "sp.pl.masked_dotplot(adata_00, selector)", but error: 'Series' object has no attribute 'split'
3. Can you guide me on how to do an advanced evaluation for this case on multiple rounds of selection (I did an advanced evaluation for only one round of selection).

![Image](https://github.com/user-attachments/assets/6dfa38ce-e6a9-46ad-826f-982e558b8f36)
![Image](https://github.com/user-attachments/assets/6572f965-dabe-4b80-aefa-7ff115be208e)
![Image](https://github.com/user-attachments/assets/48ec0ef8-2bc6-4bb0-b40c-74fb3e171c97)
Then I used "selector + selector2.plot_gene_overlap()" to get ![Image](https://github.com/user-attachments/assets/235f8cbd-0f01-4b12-a395-9a587fb65190)

**Comment by KunHHE** [comment]
Hi @LouisK92, Can I double check with you? I finished 2 rounds of selection, I targeted 300 total selections (pre-signed 140 genes, the first selected 120 genes, the second selection put pre-signed 140 genes, then it should be total 180), then I used:
1. #set the indeces of selector.probeset and adata.var to the gene symbols
selector2.probeset.index = adata_00.var.loc[selector2.probeset.index, 'feature_name'].values
2. # Concatenate the two selections
selector2.probeset['gene_nr'] += n_genes_first_selection
probeset = pd.concat([selector.probeset.loc[selector.probeset["selection"]], selector2.probeset])
3. second_selection = selector2.probeset[selector2.probeset["selection"]].index.tolist()
final_selection = first_selection + second_selection
print(len(np.unique(final_selection)))
I got 266 genes;
4. Further # Print number of selected genes in both selections
num_genes_selector = len(selector.probeset[selector.probeset["selection"]])
num_genes_selector2 = len(selector2.probeset[selector2.probeset["selection"]])
print(f"Number of genes selected from selector: {num_genes_selector}")
print(f"Number of genes selected from selector2: {num_genes_selector2}")

I got Number of genes selected from selector: 120
Number of genes selected from selector2: 172
5. I wen to "probset.csv" files from 2 separate folders, if only count "selection" columns, one is 120, one is 180.

did my script wrong to make it confused on the genes number counting?

Thanks so so much!!!
